.github/workflows/protobuf.yml

```diff
@@ -0,0 +1,42 @@
+---
+name: Protobuf specs
+on: [pull_request]
+jobs:
+  protobuf-fresh:
+    name: Compile protobuf specs to rust code
+    timeout-minutes: 30
+    runs-on: buildjet-16vcpu-ubuntu-2004
+    # runs-on: ubuntu-latest
+    steps:
+      - name: Checkout the source code
+        uses: actions/checkout@v2
+      - name: Install Rust toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: stable
+          override: false
+
+      - name: Install protobuf compiler
+        shell: bash
+        run: |
+          curl -sSfL -O https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_VERSION}/protoc-${PROTOC_VERSION}-linux-x86_64.zip
+          unzip protoc-${PROTOC_VERSION}-linux-x86_64.zip -d $HOME/.local
+        env:
+          # N.B. the major version is omitted in the download URLs,
+          # so "21.8" actually means "3.21.8".
+          PROTOC_VERSION: "21.8"
+
+      - name: Compile protobuf specs into rust src files
+        # use base for 'set -eo pipefail'
+        shell: bash
+        run: |
+          cd tools/proto-compiler
+          cargo run
+          if ! git diff --quiet; then
+              echo "ERROR: protobuf files must be regenerated and committed:"
+              git diff --stat
+              exit 1
+          else
+              echo "OK: no changes required to protobuf specs"
+          fi
```

Comment:
This will fail not just if the protos haven't been regenerated, but also if there's any changes to the output produced by `prost`, regardless if that output is meaningfully different (e.g., if a new minor version changed the whitespace).  This could lead to a bad situation where the build fails because of slightly different versions between the PR author and the CI server.

I'm not sure how likely this is, or, even if it happened, how annoying it would be, but it's something that we should watch out for.  (It's definitely good to try to catch the more common case, where someone forgets to rerun the proto compiler after editing protos).

Reply:
Great point, I hadn't realized that. Will add a descriptive comment in the workflow declaration explaining the trade-offs.

Reply:
The exact version of prost should be the same between CI and dev machine, because we have a `Cargo.lock` file, right? I suspect that we might possibly see a CI failure on a minor version bump of prost, but I don't think it will happen spuriously, it should happen when the `Cargo.lock` is updated by `cargo update`.

Conjecture: I also suspect that prost's code generation isn't likely to change spacing, because if it's printing a Rust syntax tree, that should be more stable than prost itself, since that would come from a crate like `syn` that provides parsing/printing of the Rust AST.

If it turns out that this produces a lot of false positives, we can revisit the check, but until then, I think it's very good to have.

One alternative which could be better, if there are lots of false positives: could we run the same check as a precommit hook?

Reply:
In my experience precommit hooks help individual developers to maintain baseline conformance, but aren't as useful for teams, since they're opt-in. I'd very much prefer the check to live in CI, where no one need think of it, until it helpfully nags. Should it begin to nag unhelpfully, then I hereby volunteer as tribute to take a closer look.

Reply:
sgtm!